### PR TITLE
feat: Replace network header types from bindings.rs to crate

### PR DESCRIPTION
**Pull request by maheshrayas**
fixes: https://github.com/kubernetes-sigs/blixt/issues/99

### CODE REVIEW

File: dataplane/ebpf/src/egress/icmp.rs [hunk 1]
```diff
@@ -8,42 +8,43 @@ use core::mem;
...
 };
 
-const ICMP_HDR_LEN: usize = mem::size_of::<icmphdr>();
+
 const ICMP_PROTO_TYPE_UNREACH: u8 = 3;
 
 pub fn handle_icmp_egress(ctx: TcContext) -> Result<i32, i64> {
-    let ip_hdr: *mut iphdr = unsafe { ptr_at(&ctx, ETH_HDR_LEN) }?;
-
-    let icmp_header_offset = ETH_HDR_LEN + IP_HDR_LEN;
```
Comment (shaneutt) [inline dataplane/ebpf/src/egress/icmp.rs:24]: I've noticed throughout this PR that these "offset" variables are being removed in favor of inline addition. This is a somewhat minor thing but I think it subtly reduces the readability of the code. For some of us who write a lot of network code seeing the eth header length and ipv4 header length added together is intuitive, but for some having a clear named variable to describe mnemonically the meaning of the value can be a small, but helpful thing. My suggestion would be to put these variables back:

```suggestion
    let icmp_header_offset = EthHdr::LEN + Ipv4Hdr::LEN;
```

But let me know your thoughts?

File: dataplane/ebpf/src/main.rs [hunk 2]
```diff
@@ -11,25 +11,29 @@ SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
...
-use utils::{ETH_HDR_LEN, ETH_P_IP, IPPROTO_ICMP, IPPROTO_TCP, IPPROTO_UDP};
+
+use network_types::{
+    eth::{EthHdr, EtherType},
+    ip::{Ipv4Hdr, IpProto},
+};
+use utils::ptr_at;
+
+
+//use utils::{ETH_HDR_LEN, ETH_P_IP, IPPROTO_ICMP, IPPROTO_TCP, IPPROTO_UDP};
```
Comment (shaneutt) [inline dataplane/ebpf/src/main.rs:36]: Looks like this is a leftover we can remove?

```suggestion
```

File: dataplane/ebpf/src/main.rs [hunk 3]
```diff
@@ -64,24 +68,20 @@ pub fn tc_ingress(ctx: TcContext) -> i32 {
...
-    match protocol {
-        IPPROTO_TCP => handle_tcp_ingress(ctx),
-        IPPROTO_UDP => handle_udp_ingress(ctx),
-        _ => Ok(TC_ACT_PIPE),
+    let eth_hdr: *const EthHdr = unsafe { ptr_at(&ctx, 0) }?;
+    match unsafe { *eth_hdr }.ether_type {
+        EtherType::Ipv4 => {
+            let ipv4hdr: *const Ipv4Hdr = unsafe { ptr_at(&ctx, EthHdr::LEN)? };
+            match unsafe { *ipv4hdr }.proto  {
+                IpProto::Tcp =>  handle_tcp_ingress(ctx),
```
Comment (shaneutt) [inline dataplane/ebpf/src/main.rs:76]: extra blank space:

```suggestion
                IpProto::Tcp => handle_tcp_ingress(ctx),
```

File: dataplane/ebpf/src/main.rs [hunk 4]
```diff
@@ -100,23 +100,18 @@ pub fn tc_egress(ctx: TcContext) -> i32 {
...
-        .load::<u8>(ETH_HDR_LEN + offset_of!(iphdr, protocol))
-        .map_err(|_| TC_ACT_PIPE)?;
-
-    match protocol {
-        IPPROTO_ICMP => handle_icmp_egress(ctx),
-        IPPROTO_TCP => handle_tcp_egress(ctx),
-        _ => Ok(TC_ACT_PIPE),
+    let eth_hdr: *const EthHdr = unsafe { ptr_at(&ctx, 0) }?;
+    match unsafe { *eth_hdr }.ether_type {
+        EtherType::Ipv4 => {
```
Comment (shaneutt) [inline dataplane/ebpf/src/main.rs:105]: Really glad we have the `EtherType` enum now it's very convenient :)

File: dataplane/ebpf/src/utils.rs [hunk 5]
```diff
@@ -5,23 +5,19 @@ SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
 */
 
 use core::mem;
+use network_types::{ip::Ipv4Hdr, eth::EthHdr};
 
 use aya_bpf::{bindings::TC_ACT_OK, programs::TcContext};
 
-use crate::bindings::{ethhdr, iphdr};
+
+
```
Comment (shaneutt) [inline dataplane/ebpf/src/utils.rs:13]: extra blank space:

```suggestion
```

File: dataplane/ebpf/src/utils.rs [hunk 6]
```diff
@@ -5,23 +5,19 @@ SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
...
-pub const ETH_P_IP: u16 = 0x0800;
-
-pub const IPPROTO_TCP: u8 = 6;
-pub const IPPROTO_UDP: u8 = 17;
-pub const IPPROTO_ICMP: u8 = 1;
-
-pub const ETH_HDR_LEN: usize = mem::size_of::<ethhdr>();
-pub const IP_HDR_LEN: usize = mem::size_of::<iphdr>();
+pub const ETH_HDR_LEN: usize = mem::size_of::<EthHdr>();
+pub const IP_HDR_LEN: usize = mem::size_of::<Ipv4Hdr>();
```
Comment (shaneutt) [inline dataplane/ebpf/src/utils.rs:20]: So at this point can't we just fully get rid of these constants, and expect all code to use `network_types::eth::EthHdr::LEN` and `network_types::ip::Ipv4Hdr::LEN` instead?

```suggestion
```

Or am I missing something?

**Review by shaneutt — CHANGES_REQUESTED** [review]
This is super cool that you jumped in here to take care of this, thank you!

Code looks great, I have only a few minor comments.

/ok-to-test

/cc @astoycos for review

and /cc @vadorovsky just for awareness that we're going to start using your crate, thanks for creating and maintaining it! :vulcan_salute:

**Comment by astoycos** [comment]
Thanks @maheshrayas! I will also take a look today 👍

File: dataplane/ebpf/src/egress/icmp.rs (hunk 1 above)
Comment (maheshrayas) [reply dataplane/ebpf/src/egress/icmp.rs:24]: makes sense. Updated

**Review by shaneutt — APPROVED** [review]
Looks awesome!

/ok-to-test
/approve

T hank you so much for jumping in and taking care of this one. Approved, but I'll leave the final LGTM to @astoycos.

File: dataplane/ebpf/Cargo.toml [hunk 7]
```diff
@@ -8,7 +8,8 @@ publish = false
 aya-bpf = { git = "https://github.com/aya-rs/aya", branch = "main" }
 aya-log-ebpf = { git = "https://github.com/aya-rs/aya", branch = "main" }
 common = { path = "../common" }
-memoffset = "0.8"
+memoffset = "0.9"
+network-types = { git = "https://github.com/vadorovsky/network-types", rev = "fd4ecb4" }
```
Comment (astoycos) [inline dataplane/ebpf/Cargo.toml:12]: You should just be able to use the released version here right? 
`network-types = "0.0.4"`

https://crates.io/crates/network-types

File: dataplane/ebpf/src/utils.rs [hunk 8]
```diff
@@ -5,24 +5,10 @@ SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
 */
 
 use core::mem;
+use network_types::{eth::EthHdr, ip::Ipv4Hdr};
```
Comment (astoycos) [inline dataplane/ebpf/src/utils.rs:8]: nit: Is this import needed here anymore?

**Review by astoycos — CHANGES_REQUESTED** [review]
Looking really good @maheshrayas Thanks for taking care of this 🥳 

Some small nits/fix also please remove all the xtask machinery  we have at https://github.com/kubernetes-sigs/blixt/blob/main/dataplane/xtask/src/main.rs#L29 concerning codegen!!

File: dataplane/ebpf/Cargo.toml (hunk 7 above)
Comment (vadorovsky) [reply dataplane/ebpf/Cargo.toml:12]: Actually, I made a new release just now. I think you should be able to switch to it. https://crates.io/crates/network-types/0.0.5

Comment (shaneutt) [reply dataplane/ebpf/Cargo.toml:12]: See: https://github.com/kubernetes-sigs/blixt/pull/147#discussion_r1409305695

tldr; we're pinning a SHA (as there's some stuff in `main` we need)

**Review by vadorovsky — CHANGES_REQUESTED** [review]
Thanks for pinging me. I made a new release of network-types, so you don't need to use git https://crates.io/crates/network-types/0.0.5

Besides, I think this xtask subcommand can be removed now :)

https://github.com/kubernetes-sigs/blixt/blob/main/dataplane/xtask/src/codegen.rs

File: dataplane/ebpf/Cargo.toml (hunk 7 above)
Comment (shaneutt) [reply dataplane/ebpf/Cargo.toml:12]: Oh cool, and now @vadorovsky dropped a new release anyway (thanks!):

https://github.com/kubernetes-sigs/blixt/pull/147#pullrequestreview-1757728964

So we can just use that. :tada:

**Comment by shaneutt** [comment]
> Thanks for pinging me. I made a new release of network-types, so you don't need to use git https://crates.io/crates/network-types/0.0.5
> 
> Besides, I think this xtask subcommand can be removed now :)
> 
> https://github.com/kubernetes-sigs/blixt/blob/main/dataplane/xtask/src/codegen.rs

`ACK`

Thank you! :partying_face:

**Review by astoycos — APPROVED** [review]
/lgtm
/approve

Thanks again @maheshrayas!
